### PR TITLE
(Please review carefully first) CppCheck Patch Bundle for master branch

### PR DESCRIPTION
**Pull request by Syphurith**
Seen @raven02 's recent work, i tried CppCheck to find some defects. I tried to fix some severe problems and marked those i don't know how to deal. Below is what i looked with care, and those with (mark) are just marked but not dealt. Used rules (not exactly as what it reported):
* out-of-bound reference
* statements should return but without return
* missing breaks in cases (mark)
* assert with requested effect (mark)
* `std::remove_if` (mark)
* if (unsigned_var < 0) (mark)
* never executed else-if conditions due to same condition above (mark)
* some quite weird/strange line i can not understand, dealt.
* Object would be immediately destroyed (mark)
* remove same value aside the `||`

Even i've already built one for these commits succussfully, please review with care.
**Still trying to implement a custom EmulationDir, so please wait until i changed those please.**

### CODE REVIEW

File: rpcs3/Emu/Cell/SPUThread.h [hunk 1]
```diff
@@ -258,6 +258,7 @@ struct spu_channel_4_t
 
 		sync_var.atomic_op([&out_result, &out_value, &out_count, last_value](sync_var_t& data)
 		{
+			// TODO:: Syphurith: I don't know what you tried to express, exactly.
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/SPUThread.h]: Assign bool result of comparison (data.count != 0) to out_result and use it in if() simultaneously, nothing should be wrong with it.

Comment (Syphurith) [inline rpcs3/Emu/Cell/SPUThread.h]: @Nekotekina Oh well. Then should i classify it with `()`, before remove the comment?

File: rpcs3/Emu/SysCalls/Modules/cellDmux.cpp [hunk 2]
```diff
@@ -148,7 +148,7 @@ void ElementaryStream::push_au(u32 size, u64 dts, u64 pts, u64 userdata, bool ra
 	u32 addr;
 	{
 		std::lock_guard<std::mutex> lock(m_mutex);
-
+		// TODO:: Syphurith: Should us convert this to a if statement? Since Asserts aren't included in Release builds.
```
Comment (Nekotekina) [inline rpcs3/Emu/SysCalls/Modules/cellDmux.cpp]: That's the purpose of asserts.

Comment (Syphurith) [inline rpcs3/Emu/SysCalls/Modules/cellDmux.cpp]: @Nekotekina So just ignore those in Release build. Got it.

**Comment by Syphurith** [comment]
If you want to see the full XML result of CppCheck then here it is: [Link](http://pan.baidu.com/s/1sjFjmlJ) with Pass: pyhp.

**Comment by Syphurith** [comment]
@Nekotekina Those two useless comments were removed. Then how about the others?

File: rpcs3/Emu/SysCalls/lv2/sys_spu.cpp [hunk 3]
```diff
@@ -508,7 +508,8 @@ s32 sys_spu_thread_group_terminate(u32 id, s32 value)
 		}
 	}
 
-	if (group->state <= SPU_THREAD_GROUP_STATUS_INITIALIZED || group->state == SPU_THREAD_GROUP_STATUS_WAITING || group->state == SPU_THREAD_GROUP_STATUS_WAITING)
+	// TODO:: Syphurith: Thank @danilaml (first) and let us talk about this mistake/typo. Have you used the wrong constant for the third condition?
+	if ((group->state <= SPU_THREAD_GROUP_STATUS_INITIALIZED) || (group->state == SPU_THREAD_GROUP_STATUS_WAITING) || (group->state == SPU_THREAD_GROUP_STATUS_WAITING))
```
Comment (Syphurith) [inline rpcs3/Emu/SysCalls/lv2/sys_spu.cpp]: I'm quite unsure about the **conditions** and the **return value below**, now.. And i would like to wait for @Nekotekina to confirm which is right.. before my changes. :smile:

File: rpcs3/Emu/Cell/SPUThread.h [hunk 4]
```diff
@@ -258,7 +258,7 @@ struct spu_channel_4_t
 
 		sync_var.atomic_op([&out_result, &out_value, &out_count, last_value](sync_var_t& data)
 		{
-			if ((out_result = data.count != 0))
+			if (out_result = (data.count != 0))
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/SPUThread.h]: I added additional braces around assignment to silence the gcc warning. No really need to change anything here.

File: rpcs3/Emu/FS/vfsStreamMemory.cpp [hunk 5]
```diff
@@ -38,6 +38,7 @@ u64 vfsStreamMemory::Write(const void* src, u64 size)
 
 u64 vfsStreamMemory::Read(void* dst, u64 size)
 {
+	// TODO:: Syphurith: Should us do anything? Since Asserts aren't included in Release builds.
```
Comment (Nekotekina) [inline rpcs3/Emu/FS/vfsStreamMemory.cpp]: Why you didn't remove this comment too?

File: rpcs3/Emu/RSX/RSXThread.cpp [hunk 6]
```diff
@@ -200,6 +200,7 @@ u32 RSXThread::OutOfArgsCount(const uint x, const u32 cmd, const u32 count, cons
 #define case_32(offset, step) \
     case_16(offset, step) \
     case_16(offset + 16*step, step)
+// TODO:: Syphurith: I'm not sure about whether this is correct or missing a break;.
```
Comment (Nekotekina) [inline rpcs3/Emu/RSX/RSXThread.cpp]: RSXThread.cpp is fine

File: rpcs3/Emu/SysCalls/Modules/cellCamera.h [hunk 7]
```diff
@@ -314,7 +314,8 @@ struct CellCameraAttributes
 	u32 LED;
 	u32 AUDIOGAIN;
 	u32 QS;
-	u32 NONZEROCOEFFS[1];
+// Syphurith: Found using cppcheck. This is originally stated as an single-element array, however references its [0] and [1]. So i changed this to 2 u32 units.
```
Comment (Nekotekina) [inline rpcs3/Emu/SysCalls/Modules/cellCamera.h]: Size "2" seems correct here.

File: rpcs3/Emu/SysCalls/Modules/cellSpurs.cpp [hunk 8]
```diff
@@ -2206,7 +2206,8 @@ s32 cellSpursCreateTasksetWithAttribute(vm::ptr<CellSpurs> spurs, vm::ptr<CellSp
 
 	if (!attr)
 	{
-		CELL_SPURS_TASK_ERROR_NULL_POINTER;
+		// Syphurith: If that is intentionally done, please remove the 'return'. Or else just remove this comment.
+		return CELL_SPURS_TASK_ERROR_NULL_POINTER;
```
Comment (Nekotekina) [inline rpcs3/Emu/SysCalls/Modules/cellSpurs.cpp]: With "return" it's clearly correct.

File: rpcs3/Emu/SysCalls/lv2/sys_spu.cpp (hunk 3 above)
Comment (Nekotekina) [inline rpcs3/Emu/SysCalls/lv2/sys_spu.cpp]: if (group->state <= SPU_THREAD_GROUP_STATUS_INITIALIZED || group->state == SPU_THREAD_GROUP_STATUS_WAITING)

This one seems correct. No need to write the same condition twice, if was obviously my mistake...

Comment (Nekotekina) [inline rpcs3/Emu/SysCalls/lv2/sys_spu.cpp]: Also I think that no need to add additional braces around comparisons too...

File: rpcs3/Emu/Cell/PPULLVMRecompiler.cpp [hunk 9]
```diff
@@ -5820,7 +5820,9 @@ void RecompilationEngine::ProcessExecutionTrace(const ExecutionTrace & execution
             }
         }
     }
-
+	// TODO:: Syphurith: It is said that just remove_if would cause some troubles.. I don't know if that would cause Memleak. From CppCheck:
+	// The return value of std::remove_if() is ignored. This function returns an iterator to the end of the range containing those elements that should be kept.
+	// Elements past new end remain valid but with unspecified values. Use the erase method of the container to delete them.
     std::remove_if(processed_execution_trace_i->second.begin(), processed_execution_trace_i->second.end(), [](const BlockEntry * b)->bool { return b->is_compiled; });
```
Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: I agree, it seems that erase() call is forgotten, but I'm not sure.

File: rpcs3/Emu/SysCalls/lv2/sys_spu.cpp (hunk 3 above)
Comment (Syphurith) [inline rpcs3/Emu/SysCalls/lv2/sys_spu.cpp]: @Nekotekina You're right, `>=` higher than `==` and these higher than `||`.
I just used braces to classify. If annoyed please tell me to remove those.

File: rpcs3/Emu/Cell/SPUThread.h (hunk 4 above)
Comment (Syphurith) [inline rpcs3/Emu/Cell/SPUThread.h]: Then, @Nekotekina should i leave it as it is now or revert the change? Thanks

File: rpcs3/Emu/FS/vfsStreamMemory.cpp (hunk 5 above)
Comment (Syphurith) [inline rpcs3/Emu/FS/vfsStreamMemory.cpp]: @Nekotekina I've forgot, so sorry.

File: rpcs3/Emu/RSX/RSXThread.cpp (hunk 6 above)
Comment (Syphurith) [inline rpcs3/Emu/RSX/RSXThread.cpp]: @Nekotekina So it doesn't need the `break;` for those cases.
Got it and i would remove the comment.

File: rpcs3/Emu/SysCalls/Modules/cellCamera.h (hunk 7 above)
Comment (Syphurith) [inline rpcs3/Emu/SysCalls/Modules/cellCamera.h]: @Nekotekina Thanks for confirmation. I would remove the comment then.

File: rpcs3/Emu/SysCalls/Modules/cellSpurs.cpp (hunk 8 above)
Comment (Syphurith) [inline rpcs3/Emu/SysCalls/Modules/cellSpurs.cpp]: OK. Confirmed and i would remove the comment.

File: rpcs3/Emu/Cell/PPULLVMRecompiler.cpp (hunk 9 above)
Comment (Syphurith) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: @Nekotekina Eh.. Now leave this as it is now? Since we all not sure about it.

**Comment by Syphurith** [comment]
After reading the cppcheck result again, i've seen `VirtualAlloc` and `mprotect` that in `PPUThread.cpp`, in function `void fill_ppu_exec_map(u32 addr, u32 size)`. The two functions both have return values that may indicts whether it is done succussfully or not.
Currently there isn't checks against their return value. Similiar happens in `stbimage.c`:3027.
I don't know whether we should **just let it crash** when allocation failed or try to catch those.
@Nekotekina I'm sorry to bother you again.. please tell me how to deal with those mentioned above.

File: rpcs3/Emu/Cell/PPULLVMRecompiler.cpp (hunk 9 above)
Comment (Syphurith) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: @Nekotekina I've tried to add `erase` call and compiled and run it with Disgaea 3.
The branch [Here](https://github.com/Syphurith/rpcs3/tree/p6-Bundle-Test). I think i should test it more with some other games/elfs.

Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: Try to ask @gopalsr83 after all, I didn't write this code...

File: rpcs3/Emu/Cell/SPUThread.h (hunk 4 above)
Comment (Nekotekina) [inline rpcs3/Emu/Cell/SPUThread.h]: if ((out_result = (data.count != 0)))

File: rpcs3/Emu/Cell/PPULLVMRecompiler.cpp (hunk 9 above)
Comment (Syphurith) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: @Nekotekina thank for your patience. I've built the binary that added `erase` for that.
However it doesn't need to have much difference.. So it won't cause leak there?
I don't know if i can see him on Github. Maybe i should contact him on IRC..

@gopalsr83 Should i add a `erase` call for that `remove_if`?

Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: It cannot cause leak, it just doesn't delete any element from an array, if that was intended.

Comment (Syphurith) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: @Nekotekina So i did not push that commit to this branch. Better leave it as it is now?

File: rpcs3/Emu/Cell/SPUThread.h (hunk 4 above)
Comment (Syphurith) [inline rpcs3/Emu/Cell/SPUThread.h]: @Nekotekina OK i would change that soon.

**Comment by Syphurith** [comment]
@Nekotekina I've used your suggestion for `SPUThread.h`.
Also, please take a look at `rpcs3/Emu/SysCalls/lv2/sys_spu.cpp` from 02c263fc19661ac1f866518db48c30697a1293a5 line 514.
```C++
	if ((group->state <= SPU_THREAD_GROUP_STATUS_INITIALIZED) || (group->state == SPU_THREAD_GROUP_STATUS_WAITING))
	{
		return CELL_EINVAL;
	}
```
Should it return `CELL_EINVAL` and not `CELL_ESTAT` there? If not so please tell me.

BTW I am thinking of merge all these commits into one when confirmed and ready to be merged, cause I was too `verbose` and used too many suggestion from you guys...

**Comment by Nekotekina** [comment]
Well, I think it can be merged since it fixed some bugs.

File: rpcs3/Crypto/sha1.cpp [hunk 10]
```diff
@@ -233,6 +233,7 @@ void sha1_update( sha1_context *ctx, const unsigned char *input, size_t ilen )
     size_t fill;
     uint32_t left;
 
+	// TODO:: Syphurith: Orz. It is said that size_t is an unsigned type..
```
Comment (tambry) [inline rpcs3/Crypto/sha1.cpp]: @Syphurith Spaces should be used here instead of tabs.

File: rpcs3/Emu/Cell/PPULLVMRecompiler.cpp [hunk 11]
```diff
@@ -5820,7 +5820,9 @@ void RecompilationEngine::ProcessExecutionTrace(const ExecutionTrace & execution
             }
         }
     }
-
+	// TODO:: Syphurith: It is said that just remove_if would cause some troubles.. I don't know if that would cause Memleak. From CppCheck:
+	// The return value of std::remove_if() is ignored. This function returns an iterator to the end of the range containing those elements that should be kept.
+	// Elements past new end remain valid but with unspecified values. Use the erase method of the container to delete them.
```
Comment (tambry) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: Same with above, space should be used (Why do some files use tabs and others use spaces?).

File: rpcs3/Crypto/sha1.cpp (hunk 10 above)
Comment (Syphurith) [inline rpcs3/Crypto/sha1.cpp]: Well it would be changed soon.

File: rpcs3/Emu/Cell/PPULLVMRecompiler.cpp (hunk 11 above)
Comment (Syphurith) [inline rpcs3/Emu/Cell/PPULLVMRecompiler.cpp]: I would fix that file, thx.

**Comment by Syphurith** [comment]
@Nekotekina If this could not be merged at last, could you please use some edits in your commits?
BTW i've seen you doing UI for SaveData, maybe the existing `SaveDataUtility.cpp` could help you a little, though it is only stubbed and just an UI. SaveData should be accessable when no game loaded.
Well hope you face no real problems.
@tambry How about the left? Tabs in mentioned files are switched to Space. I didn't try to use a tool for these cause Neko's repo is the newest, and i don't wanna cause much edits.
Eh.. I don't know who should i call for help in reviewing these codes..

**Comment by Bigpet** [comment]
Alright, I guess merging this is alright, I'll create an issue with the remaining warnings, so that we don't lose track of them. (also these random travis failures are a travesty, our build system and code structure really need some attention)
